app/src/main/cpp/SJCL.cpp

```diff
@@ -176,7 +273,96 @@ char* SJCL::decrypt(string sjcl_json, string key) {
     free(cyphertext);
     free(data);
     free(derived_key);
-//    free(food);
+
+    return ret;
+}
+
+char* SJCL::encrypt(char* plaintext, const string& key) {
+    int plaintext_len = strlen(plaintext);
+    int iv_len = 13;    // use 13 because 15-lol (with initial lol=2) is hardcoded in the decryptccm implementation
+
+    // strange iv_len calculation due to the decryptccm implementation
+    if (plaintext_len >= 1<<16) iv_len--;
+    if (plaintext_len >= 1<<24) iv_len--;
+
+    int salt_len = 12;  //can I use a random number here?
+    int iter = 1000;
+    int key_size = 256;
+    int tag_size = 64;
+    int ciphertext_allocation_multiplicator = 3;    // give generated ciphertext some backup space
+
+    int ks = key_size / 8;  // Make it bytes
+    int ts = tag_size / 8;  // Make it bytes
+    unsigned char *ciphertext;
+    unsigned char *derived_key;
+    unsigned char tag[ts];
+    unsigned char *iv = nullptr;
+    unsigned char *salt = nullptr;
+    unsigned char *additional = (unsigned char *)"";
+    char* ret = NULL;
+
+    srand(time(NULL));
+
+    //iv = (unsigned char *) get_random_string(iv_len).c_str();
+    std::string tmpiv = get_random_string(iv_len);
+    iv = (unsigned char *) tmpiv.c_str();
+    iv_len = tmpiv.length();
+
+    //salt = (unsigned char *) get_random_string(salt_len).c_str();
+    std::string tmpsalt = get_random_string(salt_len);
+    salt = (unsigned char *) tmpsalt.c_str();
+    salt_len = tmpsalt.length();
+
+    derived_key = (unsigned char*) malloc(sizeof(unsigned char) * ks);
+
+    // PBKDF2 Key derivation with SHA256 as SJCL does by default
+    PKCS5_PBKDF2_HMAC(key.c_str(), key.length(), salt, salt_len, iter, EVP_sha256(), ks, derived_key);
+
+    // Assuming ciphertext will not be bigger that the plaintext length * ciphertext_allocation_multiplicator
+    ciphertext = (unsigned char *) malloc(sizeof(unsigned char) * strlen(plaintext) * ciphertext_allocation_multiplicator);
+
+    // Ensure ciphertext ends up null terminated (do I need this?)
+    for (int i = 0; i < (sizeof(unsigned char) * strlen(plaintext) * ciphertext_allocation_multiplicator); i++) ciphertext[i] = '\0';
```

Comment:
In your case you only need to ensure the plaintext is null terminated, i don't think this part is necessary since all operations you make with the cyphertext are based on a known length and not trusting the string null terminator.

Reply:
Right. The line will be removed with the upcoming commit 624771ab26474a893230bcb47cdaac23932f2ff1.